x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Replace.java

```diff
@@ -0,0 +1,137 @@
+        this.regex = regex;
+        this.newStr = newStr;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataTypes.KEYWORD;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isString(str, sourceText(), FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        resolution = isString(regex, sourceText(), SECOND);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        return isString(newStr, sourceText(), THIRD);
+    }
+
+    @Override
+    public boolean foldable() {
+        return str.foldable() && regex.foldable() && newStr.foldable();
+    }
+
+    @Override
+    public Object fold() {
+        return EvaluatorMapper.super.fold();
+    }
+
+    @Evaluator(extraName = "Constant", warnExceptions = PatternSyntaxException.class)
+    static BytesRef process(BytesRef str, @Fixed Pattern regex, BytesRef newStr) {
+        if (str == null || regex == null || newStr == null) {
+            return null;
+        }
+        return new BytesRef(regex.matcher(str.utf8ToString()).replaceAll(newStr.utf8ToString()));
+    }
+
+    @Evaluator(warnExceptions = PatternSyntaxException.class)
+    static BytesRef process(BytesRef str, BytesRef regex, BytesRef newStr) {
+        if (str == null) {
+            return null;
+        }
+
+        if (regex == null || newStr == null) {
+            return str;
+        }
+        return new BytesRef(str.utf8ToString().replaceAll(regex.utf8ToString(), newStr.utf8ToString()));
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new Replace(source(), newChildren.get(0), newChildren.get(1), newChildren.get(2));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, Replace::new, str, regex, newStr);
+    }
+
+    @Override
+    public ScriptTemplate asScript() {
+        throw new UnsupportedOperationException("functions do not support scripting");
+    }
+
+    @Override
+    public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        var strEval = toEvaluator.apply(str);
+        var newStrEval = toEvaluator.apply(newStr);
+
+        if (regex.foldable() && regex.dataType() == DataTypes.KEYWORD) {
+            Pattern regexPattern;
+            try {
+                regexPattern = Pattern.compile(((BytesRef) regex.fold()).utf8ToString());
+            } catch (PatternSyntaxException pse) {
+                // TODO this is not right (inconsistent). See also https://github.com/elastic/elasticsearch/issues/100038
+                // this should generate a header warning and return null (as do the rest of this functionality in evaluators),
```

Comment:
IMO throwing here (as it now happens) is what we should align to: the argument (the reg. expression) isn't correct and we should inform the user before execution; less so, but somewhat similar to a syntax error.
We could let that happen maybe even earlier (like when folding). So the try-catch could be dropped, IMO. 
(But we can go ahead with merging this as is and revisiting when deciding on #100038).

Reply:
I am letting this one as is for now. I reckon we'll have a discussion about this general approach (not only here) and whatever we decide then we'll be applied everywhere. I don't like the inconsistent behavior, but I do also see a reasoning in throwing an exception here. I'm very much in favor of the consistency approach, though.